modules/javafx.graphics/src/main/java/com/sun/javafx/util/DataURI.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.util;
+
+import java.net.URLDecoder;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class DataURI {
+
+    /**
+     * Determines whether the specified URI uses the "data" scheme.
+     */
+    public static boolean matchScheme(String uri) {
+        if (uri == null || uri.length() < 6) {
+            return false;
+        }
+
+        int firstNonWhitespace = 0, length = uri.length();
```

Comment:
I implemented it in this way so that this requirement is not imposed onto the caller, similar to `java.net.URL` does not impose this requirement onto its callers. I can imagine that `DataURI` might be used in other places, so it might make sense to make it more robust with regards to leading whitespace?

Reply:
In that case, it might be clearer and simpler to just call `trim()` on the input String before doing anything with it, unless there is a reason not to.

Reply:
My idea was to avoid repeatedly allocating a new String with the data contained in the URI. Since `matchScheme` is called at least twice (maybe more), that would be a total of at least three copies of the URI data until we parse it. Granted, this would only apply for URIs that contain leading or trailing whitespace (since `trim()` is specified to return the same String instance if it contains no leading or trailing whitespace).

Reply:
I see. How about something like this then?

```
    if (uri == null || uri.isEmpty()) {
        return false;
    }

    if (Character.isWhiteSpace(uri.charAt(0)) {
        uri = uri.trim();
    }

    if (uri.length() < 6) {
        return false;
    }
```

The rest of the method can then work on a trimmed string.

Reply:
I replaced it with another implementation that is effectively the same as this.